api/buckets/service.go

```diff
@@ -412,23 +412,37 @@ func (nn *namedNodes) Store(c cid.Cid, n *namedNode) {
 // This method returns a map of all nodes keyed by their _original_ plaintext cid.
 func (s *Service) encryptDag(ctx context.Context, ds ipld.DAGService, root ipld.Node, key []byte, add *namedNode) (map[cid.Cid]*namedNode, error) {
 	// Step 1: Create a preordered list of joint and leaf nodes
-	var stack, joints, leaves []*namedNode
+	var stack, joints []*namedNode
+	jmap := make(map[cid.Cid]*namedNode)
+	lmap := make(map[cid.Cid]*namedNode)
 	stack = append(stack, &namedNode{node: root})
 	var cur *namedNode
 
 	for len(stack) > 0 {
 		n := len(stack) - 1
 		cur = stack[n]
 		stack = stack[:n]
+
+		if _, ok := jmap[cur.node.Cid()]; ok {
```

Comment:
Don't traverse the node if we already have traversed the exact same on elsewhere.